Feature/aidumov/dp 4 registration screen

superaydumov opened the pull request:
PR по выполненной задаче DP-4 по созданию экрана регистрации пользователя.
Добавлен координатор для управления навигацией.

superaydumov committed:
[DP-4] - Реорганизовал файлы в проекте (разложил по папкам).

superaydumov committed:
[DP-4] - Сверстал экран авторизации пользователя, добавил появление предупреждений в соответствии с макетом.

superaydumov committed:
[DP-4] - Изменил вызов AuthViewController из OnboardingViewController.

superaydumov committed:
[DP-4] - Добавил coordinator, presenter и sceneFactory к флоу аутентификации authFlow (после окончания авторизации идет вызов mainFlow).
В функции showMainFlow класса AppCoordinator используется заглушка для показа MainViewController.

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-4-Registration-screen

superaydumov committed:
[DP-4] - Исправлены отступы для соответствия установленным правилам.

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthCoordinator.swift:
```diff
@@ -0,0 +1,26 @@
+    }
+}
+
+private extension AuthCoordinator {
+    func showAuthentification() {
```
Ошибка нейминга - пустая строка после `extension`

superaydumov replied:
Fixed

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthPresenter.swift:
```diff
@@ -0,0 +1,29 @@
+
+import Foundation
+
+protocol AuthPresenterProtocol: AnyObject {
+    func authFinish()
+}
```
Протоколы в отдельном файле или в одном? `AuthViewProtocol` в отдельном, а `AuthPresenterProtocol` в том же... 🤷🏻‍♂️

superaydumov replied:
Согласен. Вынес в отдельный файл

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+    private enum Keys {
+        static let upperLabelText = "Введите ваше имя"
+        static let lowerLabelInputNameWarningText = "Введите имя"
+        static let lowerLabelLengthWarhingText = "Минимум две буквы"
```
Опечатка

superaydumov replied:
Fixed

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+        textField.backgroundColor = .clear
+        textField.font = .textHeadingFont
+        textField.textAlignment = .center
+        textField.becomeFirstResponder()
```
Если `textField.becomeFirstResponder()` запустить в `viewDidLoad`, то клавиатура сама появится и не нужен будет NotificationObserver

superaydumov replied:
Если убрать observer, то кнопку "войти" необходимо будет привязывать вручную.
Сейчас это происходит автоматически при сигнале на появление от клавиатуры.
becomeFirstResponder() просто нам её показывает. Находясь внутри ленивого свойства, он по сути тоже активируется при построении экрана авторизации.

superaydumov replied:
Изменил код по твоему совету с textField.inputAccessoryView = enterButton.
Удалил логику с observer

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+    // MARK: - Private methods
+
+    private func setupSubviews() {
+        [upperLabel,
```
В этом случае скобка массива `[` на пустой строке

superaydumov replied:
Fixed

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+
+    private func hideLowerLabel() {
+        lowerLabel.text = ""
+        lowerLabelHeightAnchor.constant = .zero
```
попробуй как альтернативу: `lowerLabel.isHidden = true`

Но решать это должен презентер

superaydumov replied:
Изменил логику работы появления lowerLabel.
Убрал лишнее свойство lowerLabelHeightAnchor

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+    private func showLowerLabel(with text: String) {
+        lowerLabel.text = text
+        lowerLabelHeightAnchor.constant = 16
+    }
```
Решать что показывать и как показывать - это функция презентера

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+        } else {
+            nameTextField.font = .textHeadingFont
+            hideLowerLabel()
+        }
```
Это функция презентера

superaydumov replied:
Перенес функцию в презентер

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+
+        // TODO: - add code to pass user name to MainViewController and save it on server
+
+        DispatchQueue.main.async {
```
Почему сюда не передается Имя пользователя?
То что пока некуда его передавать - не основание не принести его сюда )

superaydumov replied:
Fixed

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+    // MARK: - Stored properties
+
+    var presenter: AuthPresenter?
+
+    private var lowerLabelHeightAnchor = NSLayoutConstraint()
```
Рекомендую сделать переменную для хранения имени пользователя. При вводе использовать его как текущее состояние имени, при нажатии кнопки - использовать его как финальное имя для передачи на следующий экран. 
Тогда этот VC можно сделать переиспользуемым - в инит передавать имя для его редактирования (для флоу редактирования имени) или по дефолту использовать пустую строку

superaydumov replied:
Планировал позже это сделать, но ладно.
Добавил свойство для имени пользователя, обновил инициализатор.

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+        replacementString string: String
+    ) -> Bool {
+        let maximumLength = charactersMaxNumber
+        let currentstring = (nameTextField.text ?? "") as NSString
```
Зачем тут `NSString`?

И еще тут опечатка

superaydumov replied:
Опечатку исправил.

NSString тут потому, что мы пользуемся методом replacingCharacters(in:with:), который работает только с типом NSString. Иначе нужно конвертировать String в NSString.

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+        let maximumLength = charactersMaxNumber
+        let currentstring = (nameTextField.text ?? "") as NSString
+        let updatedString = currentstring.replacingCharacters(in: range, with: string)
+        if updatedString.rangeOfCharacter(from: CharacterSet.decimalDigits) != nil {
```
Нужно добавить проверку на наличие пробела и символов ИЛИ показывать другой ворнинг, не "Только буквы"

superaydumov replied:
Fixed

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+}
+    // MARK: - UITextFieldDelegate
+
+extension AuthViewController: UITextFieldDelegate {
```
Если нужно чтобы нажатие на клавиатуре было идентично нажатию экранной кнопки "войти", тогда нужно добавить еще одну функцию делегата
```swift
func textFieldShouldReturn(_ textField: UITextField) -> Bool { }
```

superaydumov replied:
Добавил, но подождем что скажет заказчик.

superaydumov replied:
Заказчик сказал, что было бы неплохо иметь такой функционал

SKemenov replied:
Классно, что сделал!

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewProtocol.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+protocol AuthViewProtocol: AnyObject {
+    // TODO: - add code to use viewControllers method in presenter
```
Логично в протокол добавить функции `Показать ворнинг`, `Скрыть ворнинг`

superaydumov replied:
Заменил функции на updateUIElements

SKemenov commented on DavaiPosmotrim/AuthFLow/AuthViewController.swift:
```diff
@@ -9,8 +9,217 @@ import UIKit
+        }
+    }
+
+    @objc func enterButtonDidTap(sender: AnyObject) {
```
Это функция презентера

superaydumov replied:
Не совсем понял
Как экшен кнопки вьюконтроллера может быть функцией презентера?

SKemenov replied:
По аналоги с методом updateUIElement

SKemenov reviewed (changes requested):
ИМХО излишни усложнил работу с клавиатурой, попробуй обойтись без observer.
Часть логики осталась в VC, стоит перенести её в презентер.
В паиарке внизу висит ворнинг, что есть конфликт с дев. Не забывай мерджить к себе дев, перед тем как делать пиарку.

И надо согласовать с Заказчиком:
- допустим ли ввод пробела и символов
- нажатие Ввод на клавиатуре обрабатываем или нет

superaydumov commented:
Отвечаю на главный коммент:
Метод с наблюдателем сделан для того, чтобы кнопка войти, нижний отступ которой привязан к верху клавиатуры, знала, что клавиатура появится и привязалась к ней. Так выходит более универсальный метод для всех размеров экранов, где не приходится задавать размеры для каждого случая.

Перед отправкой на проверку, я влил ветку develop в свою (выше запись об этом есть). После создания моего PR были влиты изменения связанные с главным экраном. Отсюда и предупреждение о конфликте.

Вопросы для заказчика вынесены в таблицу.

superaydumov committed:
[DP-4] - Вынес AuthPresenterProtocol в отдельный файл.
Поправил отступы в соответствии с правилами, исправил ошибки в названиях переменных.

superaydumov committed:
[DP-4] - Изменил логику работы появления lowerLabel.
Убрал лишнее свойство lowerLabelHeightAnchor

superaydumov committed:
[DP-4] - Добавил переменную для имени пользователя (обновил инициализатор класса для передачи имени извне).

superaydumov committed:
[DP-4] - Обновил логику появления предупреждения "Только буквы" при введении цифр, знаков и пробелов (разрешены только буквы).
Добавил функцию updateUIElements вместо showLowerLabel и hideLowerLabel.

superaydumov committed:
[DP-4] - Исправил отступы

superaydumov committed:
[DP-4] - Исправление увеличения шрифта нижнего лейбла при нажатии на кнопку символа, цифры, пробела.

superaydumov committed:
[DP-4] - Заменил свойство в методе shouldChangeCharactersIn.

superaydumov committed:
[DP-4] - Перенес бизнес логику подсчета количества символов в строке в презентер.
В презентере происходит вызов функции обновления UI элементов вьюконтроллера.
Добавил enum Resources  для хранения строк в одном месте.

superaydumov committed:
[DP-1] - Перенес функцию updateUIElements в расширение AuthViewController.

superaydumov committed:
Revert "[DP-1] - Перенес функцию updateUIElements в расширение AuthViewController."

This reverts commit aed40989e134b86449ee1760332d9b879a00cf78.

superaydumov committed:
[DP-4] - Перенес функцию updateUIElements в расширение AuthViewController.

superaydumov committed:
Merge branch 'develop' into feature/aidumov/DP-4-Registration-screen

# Conflicts:
#	DavaiPosmotrim.xcodeproj/project.pbxproj
#	DavaiPosmotrim/Application/AppCoordinator.swift

superaydumov committed:
[DP-4] - Добавил сохранение имени пользователя в UserDefaults.
Обновил логику работы потоков с учетом сохраненного имени.
Исправил отступы.

superaydumov committed:
[DP-4] - Добавил показ предупреждения при достижении максимального количества символов.

superaydumov committed:
[DP-4] - Изменил привязку кнопки "Войти" на экране (сейчас связь идет через textField.inputAccessoryView = enterButton).
Убрал логику с observer.

Sbat2000 commented on DavaiPosmotrim/AuthFLow/AuthSceneFactory.swift:
```diff
@@ -0,0 +1,18 @@
+
+struct AuthSceneFactory {
+    static func makeAuthViewController(with coordinator: AuthCoordinator) -> AuthViewController {
+        let savedUserName = UserDefaults.standard.string(forKey: "userName") ?? ""
```
Я бы проверял имя в презентере после загрузки View, по мне так это более логично.

superaydumov replied:
Добавил в презентер функцию по проверке сохраненного имени.
Вызов происходит в этом месте перед передачей строки во вьюконтроллер.

Sbat2000 replied:
Привет! я имел ввиду чуть другое: координатор по идее не должен вообще ничего знать об имени. Я вижу логику так: наше вью загружается, говорит презентеру, что оно загружено, презентер в этот момент должен провреить имя в ЮД и отдать его вьюконтроллеру. Или ничего не отдать, если его нет.

superaydumov replied:
Fixed
Перенес логику вызова сохраненного имени в AuthViewController

Sbat2000 commented on DavaiPosmotrim/AuthFLow/AuthCoordinator.swift:
```diff
@@ -0,0 +1,29 @@
+
+    func showAuthentication() {
+        let viewController = AuthSceneFactory.makeAuthViewController(with: self)
+        navigationController.pushViewController(viewController, animated: true)
```
Вот тут надо понять стоит ли нам пушить в текущий стек навигации или лучше обнулить его и туда рутом поставить только наш ВК авторизации? мне кажется второй вариант лучше, так как мы не должны дать пользователю нажать назад? а значит нам не нужны старые контроллеры (онбординг) в стеке.

superaydumov replied:
Согласен))
Использовал `navigationController.setViewControllers([viewController], animated: true)` для обнуления стека навигации и назначения вьюконтроллера авторизации корневым

SKemenov reviewed (approved):
Классно, что смог упростить код. Отличная идея с `Resources` 👍

superaydumov committed:
[DP-4] - Исправление замечаний:
- вынес бизнес-логику кнопки "Войти" в презентер (метод handleEnterButtonTap);
- создал в презентере функцию по проверке сохраненного имени пользователя (метод checkUserNameProperty);
- стек навигации обнуляется при запуске AuthCoordinator.

superaydumov committed:
[DP-4] - Исправил отступ.

superaydumov committed:
[DP-4] - Перенес логику вызова сохраненного имени внутрь AuthViewController.
Имя будет запрашиваться каждый раз при вызове экрана авторизации.

superaydumov committed:
[DP-4] - Дополнил файл Resources ключом для сохранения имени

superaydumov committed:
[DP-4] - Отрефакторил метод updateUIElements.
Исправил отступы в методе checkUserNameProperty.

uistelmakh commented on DavaiPosmotrim/Application/AppCoordinator.swift:
```diff
@@ -34,8 +36,12 @@ private extension AppCoordinator {
+            type: .auth,
+            finishDelegate: self,
+            navigationController: navigationController)
+        addChild(authCoordinator)
```
не нужна объект создавать authCoordinator. Сразу добавляй в метод
addChild(
           AuthCoordinator(
            type: .auth,
            finishDelegate: self,
            navigationController: navigationController
            )
         )

superaydumov replied:
А как мне тогда его стартовать?
Сейчас запускается так: authCoordinator.start()

uistelmakh replied:
а все окей

uistelmakh commented on DavaiPosmotrim/AuthFLow/AuthCoordinator.swift:
```diff
@@ -0,0 +1,29 @@
+    override func finish() {
+        finishDelegate?.didFinish(self)
+    }
+
```
лишний отступ, в экстеншне тоже

superaydumov replied:
Fixed.
Также исправил отступы в файле Readme этой ветки.

uistelmakh commented on DavaiPosmotrim/AuthFLow/AuthPresenter.swift:
```diff
@@ -0,0 +1,94 @@
+            return ""
+        } else {
+            UserDefaults.standard.setValue(name, forKey: Resources.Authentication.savedNameUserDefaultsKey)
+            DispatchQueue.main.async {
```
в презентере не обновляй UI. Отображением занимается View слой

superaydumov replied:
Fixed

superaydumov committed:
[DP-4] - Исправил отступы.

superaydumov committed:
Update README.md

superaydumov committed:
[DP-4] - Исправил замечания после ревью

superaydumov committed:
[DP-4] - Добавил отступ
